pkg/controllers/state/metrics/node.go

```diff
@@ -0,0 +1,251 @@
+	overheadGaugeVec       = newNodeGaugeVec("system_overhead", "Node system daemon overhead are the resources reserved for system overhead, the difference between the node's capacity and allocatable values are reported by the status. Labeled by provisioner name, node name, zone, architecture, capacity type, instance type, node phase and resource type.")
+)
+
+func newNodeGaugeVec(name, helpMessage string) *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      name,
+			Help:      helpMessage,
+		},
+		nodeLabelNames(),
+	)
+}
+
+func nodeLabelNames() []string {
+	labels := []string{
+		resourceType,
+		nodeName,
+		nodeProvisioner,
+		nodePhase, // NOTE: deprecated
+	}
+
+	for _, l := range getWellKnownLabels() {
+		labels = append(labels, l)
+	}
+
+	return labels
+}
+
+type nodeScraper struct {
+	cluster  *state.Cluster
+	labelMap map[string]map[*prometheus.GaugeVec][]prometheus.Labels
+}
+
+func newNodeCollector(cluster *state.Cluster) *nodeScraper {
+	return &nodeScraper{
+		cluster:  cluster,
+		labelMap: make(map[string]map[*prometheus.GaugeVec][]prometheus.Labels),
+	}
+}
+
+func (ns *nodeScraper) getName() string {
+	return "nodes"
+}
+
+func (ns *nodeScraper) init(ctx context.Context) {
+	for _, gauge := range []*prometheus.GaugeVec{
+		allocatableGaugeVec,
+		podRequestsGaugeVec,
+		podLimitsGaugeVec,
+		daemonRequestsGaugeVec,
+		daemonLimitsGaugeVec,
+		overheadGaugeVec,
+	} {
+		crmetrics.Registry.Register(gauge)
+	}
+}
+
+func (ns *nodeScraper) update(ctx context.Context) {
+	nodes := make(map[string]struct{})
+	ns.cluster.ForEachNode(func(n *state.Node) bool {
+		if _, ok := ns.labelMap[n.Node.Name]; !ok {
+			logging.FromContext(ctx).Infof("Tracking new node: %s", n.Node.Name)
+			ns.labelMap[n.Node.Name] = make(map[*prometheus.GaugeVec][]prometheus.Labels)
+		}
+		nodes[n.Node.Name] = struct{}{}
+
+		podRequests := resources.Subtract(n.PodTotalRequests, n.DaemonSetRequested)
+		podLimits := resources.Subtract(n.PodTotalLimits, n.DaemonSetLimits)
+		allocatable := n.Node.Status.Capacity
+		if len(n.Node.Status.Allocatable) > 0 {
+			allocatable = n.Node.Status.Allocatable
+		}
+
+		// Populate  metrics
+		for gaugeVec, resourceList := range map[*prometheus.GaugeVec]v1.ResourceList{
+			overheadGaugeVec:       ns.getSystemOverhead(n.Node),
+			podRequestsGaugeVec:    podRequests,
+			podLimitsGaugeVec:      podLimits,
+			daemonRequestsGaugeVec: n.DaemonSetRequested,
+			daemonLimitsGaugeVec:   n.DaemonSetLimits,
+			allocatableGaugeVec:    allocatable,
+		} {
+			if err := ns.set(resourceList, n.Node, gaugeVec); err != nil {
+				logging.FromContext(ctx).Errorf("Failed to generate gauge: %s", err)
+			}
+		}
+
+		return true
```

Comment:
I'm not sure. I don't know if failing to generate one gauge should mean we shouldn't try to generate the rest of them. However, I'm not sure what scenarios  can cause generation to fail. If one failing means all following will fail it would make sense to exit early.

Reply:
Ah I see, so if this returns false, we wouldn't generate gauges for other nodes. Sounds good to me.